### DISCUSSION TITLE
add option to `var()` for escaping invalid character when injecting into namespace

### DISCUSSION THREAD

**Discussion by kompre** [sympy/sympy#25832] [Ideas]
Hello,

the `var()` function is used to create a symbol and then inject it into the current framework as a variable with the same name. 

`Symbol` object can be define with invalid characters that would raise syntax error if try to use directly, for example:

![image](https://github.com/sympy/sympy/assets/11796345/dad42e40-2eb1-4179-a3a3-9d0dd151a6cb)

The object are valid Symbols, but can't be accessed directly

![image](https://github.com/sympy/sympy/assets/11796345/576e7bf3-251e-43dd-8581-8d807fa6e29d)

It would be nice if there would be an option to escape invalid characters with safe ones, possibly customizeable by the user (e.g. a list of regex).

As it stand right now `var(names, **kwargs)` pass the `names, **kwargs` to the `symbol()`, which returns a list of `Symbol`s (all fine). For the actual namespace injection it calls `Symbol.name` to retrieve the string representation of the symbol

```py

# current code from var()
try:
        syms = symbols(names, **args)

        if syms is not None:
            if isinstance(syms, Basic):
                frame.f_globals[syms.name] = syms # here calls syms.name
            elif isinstance(syms, FunctionClass):
                frame.f_globals[syms.__name__] = syms # here syms.__name__ for 
            else:
                traverse(syms, frame) # for list of symbols
    finally:
        del frame  # break cyclic dependencies as stated in inspect docs
```

I propose to add as argument a list of substitution such as:

```py
def var(names, list_of_substitution = None, **kwargs):
    # new escape funcion
    def escape_name(symbol_name, list_of_substitution):
        name = symbol_name # should create a copy of the string, so not to change the actual symbol name
        for old, new in list_of_substitution.items():
            name = name.replace(old, new)
        return name

    #
    # .....
    #
    # call the escape function on name
    try:
        syms = symbols(names, **args)

        if syms is not None:
            if isinstance(syms, Basic):
                
                frame.f_globals[escape_name(syms.name)] = syms # here calls syms.name
            elif isinstance(syms, FunctionClass):
                frame.f_globals[escape_name(syms.__name__)] = syms # here syms.__name__ for 
            else:
                traverse(syms, frame) # for list of symbols
    finally:
        del frame  # break cyclic dependencies as stated in inspect docs

```

In my case I would like the to use the convention to substitute `{` `}` with empty space, and `,` with a `_`:

![image](https://github.com/sympy/sympy/assets/11796345/2adab053-bfb0-469a-8b37-16fd7d5a1706)

**Comment by oscarbenjamin** [comment]
I think it would be better just to use `Symbol` directly like:
```
F_d_t = Symbol('F_{d,t}')
```
Then it is clear what the variable name is rather than needing to figure out what rules `var` would have for turning arbitrary strings into valid Python identifiers.

Generally I would discourage the use of `var`.

**Reply by kompre** [reply]
I agree with you about `var` and distrusted it for a bit, but I'm looking to have cleaner code cells in my ipynb, in order to be easy to use on a whim, even for easy calculations. 

Right now I'm experimenting with `parse_expr`, where theoretically I can not even define any symbols before parsing an expression. For example:

```py
from sympy import *
from sympy.parsing.sympy_parser import parse_expr, T

ex = parse_expr(
    "F_d_t * x * c_a",
    local_dict={
        "F_d_t": Symbol("F_{d,t}"),
    },
)
ex
```
![image](https://github.com/sympy/sympy/assets/11796345/fc1c58e1-73e6-447a-a553-e28bead17566)

I got a valid expression without the need to define any `Symbol`s beforehand because I was fine with the default handling of symbols, while passing a `local_dict` option where I specified how to handle that special symbol.

If I want to inject the symbols into the namespace for easy later retrieval, I could do something like this (pseudocode):

```py
def escape_symbol_name(symbol : Symbol):
    # function that escapes character that would raise syntax errors
    return escaped_symbol_name # it's a string
    
# inject symbol in namespace    
for symbol in ex.free_symbols:
    globals()[escape_symbol_name(symbol)] = symbol 
```

Technically I could get away with defining any symbol, using just the `parse_expr` a local_dict, but it would get tiresome really fast because each time I want to use the same "special" symbol, I would need to pass the local_dict containing it, while if I inject the symbol into the namespace, I could just pass the `global_dict` to the `parse_expr`.

```py
from sympy import *
from sympy.parsing.sympy_parser import parse_expr, T

ex = parse_expr(
    "F_d_t * x * c_a",
    local_dict={
        "F_d_t": Symbol("F_{d,t}"),
    },
)
ex

# inject symbol in namespace    
for symbol in ex.free_symbols:
    escape_var(symbol) # I'm using my modified version of var that escapes invalid character, and ALSO accept Symbols as input
    
ex2 = parse_expr(
    "F_d_t + x + c_a + ex",
    global_dict=globals(), # I don't redefine `F_d_t`, I just pass the globals() dict that contains it
)
ex2
```
![image](https://github.com/sympy/sympy/assets/11796345/ff5751c9-7d66-42b0-b312-bc069bc8ab9a)

I just think the code cells looks neat this way. 

I attach the definition of the function `escape_var` which I am currently using in ipynb: I just copied the code from the `var` function and modified to escape the character, but **also** to accept `Symbol` object as input:

```py
def escape_name(symbol_name, list_of_substitution):
    name = str(symbol_name)
    for old, new in list_of_substitution.items():
        name = name.replace(old, new)
    return name

def escape_var(names, list_of_substitution=None, **args):
    
    def traverse(symbols, frame):
            """Recursively inject symbols to the global namespace. """
            for symbol in symbols:
                if isinstance(symbol, Basic):
                    frame.f_globals[escape_name(symbol.name, list_of_substitution)] = symbol # modified
                elif isinstance(symbol, FunctionClass):
                    frame.f_globals[escape_name(symbol.__name__, list_of_substitution)] = symbol # modified
                else:
                    traverse(symbol, frame)

    from inspect import currentframe
    frame = currentframe().f_back

    try:
        # modified to accept `Symbol` as valid input
        if isinstance(names, str):
            syms = symbols(names, **args) # normal var behaviour
        else:
            syms = names # accept Symbol object

        if syms is not None:
            if isinstance(syms, Basic):
                frame.f_globals[escape_name(syms.name, list_of_substitution)] = syms # modified
            elif isinstance(syms, FunctionClass):
                frame.f_globals[escape_name(syms.__name__, list_of_substitution)] = syms # modified
            else:
                traverse(syms, frame)
    finally:
        del frame  # break cyclic dependencies as stated in inspect docs

    return syms
```

**Comment by oscarbenjamin** [comment]
I think that a better approach would be to register an ipython magic command:
https://ipython.readthedocs.io/en/stable/config/custommagics.html

This is something that SymPy could do at start up somehow.

Then you should be able to do e.g.
```
@syms x, y
```
and the idea would be that this would be equivalent to something like
```
x, y = symbols('x, y')
all_symbols.extend([x, y])
```
